libs/SalesforceNetwork/SalesforceNetwork/SalesforceNetwork/Classes/Network/Action/CSFAction.m

```diff
@@ -535,21 +535,24 @@ - (void)completeOperationWithError:(NSError *)error {
 
 - (id)contentFromData:(NSData*)data fromResponse:(NSHTTPURLResponse*)response error:(NSError**)error {
     id content = nil;
-    
+    BOOL requestSucceeded = (response.statusCode >= 200 && response.statusCode < 300);
+
+    // try to parse response if response is not an error or response status code is between the specified status code range
+    // 2xx is for successful request
+    // 4xx is for client error that may contains valuable error information in response
     NSError *jsonParseError = nil;
     if ([self.responseData length] > 0) {
```

Comment:
This came from `Wave` IIRC. @qingqingliu or @jeanbovet might remember. I pulled it in from their fork and made a PR here. But I agree with the approach you're proposing. It's in line with our discussion from earlier this week. Let's do it!